model/routing/ResourceUrlBuilder.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2018 (original work) Open Assessment Technologies SA ;
+ */
+
+namespace oat\taoBackOffice\model\routing;
+
+use oat\generis\model\OntologyAwareTrait;
+use oat\oatbox\service\ConfigurableService;
+use oat\tao\model\menu\MenuService;
+use oat\tao\model\menu\Perspective;
+use oat\tao\model\menu\Section;
+
+class ResourceUrlBuilder extends ConfigurableService
+{
+    use OntologyAwareTrait;
+
+    const SERVICE_ID = 'taoBackOffice/resourceUrlBuilder';
+
+    const OPTION_CLASS_EXTENSION_ASSOCIATIONS = 'class_to_extension_associations';
+
+    /**
+     * Builds a full URL to be used by front-end for a resource based on the OPTION_CLASS_EXTENSION_ASSOCIATIONS map
+     *
+     * Add your association if it is missing.
+     *
+     * @param \core_kernel_classes_Resource $resource
+     * @return string
+     */
+    public function buildUrl(\core_kernel_classes_Resource $resource)
```

Comment:
As for me, this function is great with lots of different parts inside.
Probably, it would look better if it were divided into several smaller functions.